modulus/distributed/utils.py

```diff
@@ -472,41 +487,36 @@ def scatter_v_wrapper(
     rank = dist.get_rank(group=group)
     if len(sizes) != comm_size:
         raise ValueError()
-    if dim >= tensor.dim():
+    if dist.get_rank(group=group) == 0 and dim >= tensor.dim():
         raise ValueError()
     if not (0 <= src < comm_size):
         raise ValueError()
 
+    # all_to_all is already all_to_all_v, use empty tensors to "mask"-out irrelevant parts
     tensor_shape = list(tensor.shape)
-    tensor_shape[dim] = sizes[rank]
-    output = torch.empty(
-        tensor_shape,
-        dtype=tensor.dtype,
-        device=tensor.device,
-    )
-
-    # dist.scatter doesn't support tensors of different shape
-    # so this implementation is using explicit send/recv combinations
-    scatter_list = None
+    x_send = [None] * comm_size
+    x_recv = [None] * comm_size
     if rank == src:
         scatter_list = torch.split(tensor, sizes, dim=dim)
-        req_list = [None] * comm_size
+        scatter_list = [t.contiguous() for t in scatter_list]
+        x_send = scatter_list
+    else:
         for r in range(comm_size):
-            tensor_to_scatter_to_r = scatter_list[r]
-            if r == src:
-                output = tensor_to_scatter_to_r
-            else:
-                req_list[r] = dist.isend(tensor_to_scatter_to_r, dst=r, group=group)
+            tensor_shape[dim] = 0
+            x_send[r] = torch.empty(
+                tensor_shape, device=tensor.device, dtype=tensor.dtype
+            )
 
-        for r in range(comm_size):
-            if r != src:
-                req_list[r].wait()
+    for r in range(comm_size):
+        if r == src:
+            tensor_shape[dim] = sizes[rank]
+        else:
+            tensor_shape[dim] = 0
+        x_recv[r] = torch.empty(tensor_shape, device=tensor.device, dtype=tensor.dtype)
 
-    else:
-        req = dist.irecv(output, src=src, group=group)
-        req.wait()
+    dist.all_to_all(x_recv, x_send, group=group)
```

Comment:
we can use all_to_all (natively supporting all_to_all_v) directly to avoid manual recv/send pairs which are also a bit more prone to hangs

Reply:
This is a nice improvement. Like in the previous comment, wondering if we can also avoid the explicit `torch.split` above and the for loop over all processes and just use `dist.all_to_all_single`

Reply:
I thought about that, too. `all_to_all_single`, however, is marked as "experimental and subject to change" at the moment. Hence, I would go for the more stable version for now.

Reply:
That's the case for `all_to_all` too though: https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_to_all

Reply:
True, realized that just now. 
`all_to_all_single` however still comes with the restriction of splits only working alongside dim0. An explicit split currently would be more flexible for now. Hence, I would vote for keeping the current approach as is.

Reply:
Makes sense. We can always update this to `all_to_all_single` down the line if we feel like it might help with perf.